### PR TITLE
TaskletFusion

### PR DESCRIPTION
**Pull request by phschaad**
Adapted from https://github.com/spcl/daceml, originally by @orausch.

Fuses two Tasklets when they're only separated by one Memlet or one AccessNode, and the first Tasklet is used nowhere else.

Note, this replaces SimpleTaskletFusion.

Future improvements / current limitations:
- The transformation is a bit strict and could be applied in more settings. The example below illustrates one such setting. Tasklet `a` could semantically be fused into the remaining Tasklets, but because it has multiple outputs (and consequently multiple assignments), the transformation does not apply.

Example from the provided test case:
image
image

Co-authored-by: @orausch

### CODE REVIEW

File: dace/transformation/helpers.py [hunk 1]
```diff
@@ -927,3 +927,24 @@ def redirect_edge(state: SDFGState,
     new_edge = state.add_edge(new_src or edge.src, new_src_conn or edge.src_conn, new_dst or edge.dst, new_dst_conn
                               or edge.dst_conn, memlet)
     return new_edge
+
+
+def find_name_not_in_set(name_set: Set[str], target: Optional[str]) -> str:
```
Comment (tbennun) [inline dace/transformation/helpers.py:932]: this exists in dace.data.find_new_name

File: tests/transformations/tasklet_fusion_test.py [hunk 2]
```diff
@@ -19,60 +23,98 @@ def map_with_tasklets(A: dace.float32[20], B: dace.float32[10]):
     return C
 
 
-def _make_sdfg(l: str = 'Python'):
-
-    language = dtypes.Language.Python if l == 'Python' else dtypes.Language.CPP
```
Comment (tbennun) [inline tests/transformations/tasklet_fusion_test.py:24]: this was quite important for c2dace. What happened?

File: tests/transformations/tasklet_fusion_test.py [hunk 3]
```diff
@@ -19,60 +23,98 @@ def map_with_tasklets(A: dace.float32[20], B: dace.float32[10]):
...
 
 
-@pytest.mark.parametrize("l", [pytest.param('Python'), pytest.param('CPP')])
-def test_map_with_tasklets(l: str):
-    sdfg = _make_sdfg(l)
-    simplify_reduced = [xf for xf in simplification_transformations() if xf.__name__ != 'SimpleTaskletFusion']
+@pytest.mark.parametrize('with_data', [pytest.param(True), pytest.param(False)])
+def test_map_with_tasklets(with_data: bool):
+    sdfg = _make_sdfg(with_data)
+    sdfg.save('_dacegraphs/orig_' + ('d' if with_data else 'nd') + '.sdfg')
```
Comment (tbennun) [inline tests/transformations/tasklet_fusion_test.py:101]: why?

File: tests/transformations/tasklet_fusion_test.py [hunk 4]
```diff
@@ -19,60 +23,98 @@ def map_with_tasklets(A: dace.float32[20], B: dace.float32[10]):
...
+    sdfg.compile()
+    simplify_reduced = [xf for xf in simplification_transformations() if xf.__name__ != 'TaskletFusion']
     sdfg.apply_transformations_repeated(simplify_reduced)
-    num = sdfg.apply_transformations_repeated(SimpleTaskletFusion)
-    assert (num == 4)
+    num = sdfg.apply_transformations_repeated(TaskletFusion)
+    sdfg.save('_dacegraphs/xformed_' + ('d' if with_data else 'nd') + '.sdfg')
```
Comment (tbennun) [inline tests/transformations/tasklet_fusion_test.py:106]: same

File: tests/transformations/tasklet_fusion_test.py [hunk 5]
```diff
@@ -19,60 +23,98 @@ def map_with_tasklets(A: dace.float32[20], B: dace.float32[10]):
...
+    assert (num == 3)
     func = sdfg.compile()
-    A = np.arange(1, 21, dtype=np.float32)
-    B = np.arange(1, 11, dtype=np.float32)
-    C = np.zeros((10, ), dtype=np.float32)
+    A = np.arange(1, N + 1, dtype=np_datatype)
+    B = np.arange(1, M + 1, dtype=np_datatype)
+    C = np.zeros((M, ), dtype=np_datatype)
     func(A=A, B=B, C=C)
+    map_with_tasklets.to_sdfg().save('_dacegraphs/tmp.sdfg')
```
Comment (tbennun) [inline tests/transformations/tasklet_fusion_test.py:113]: same

File: tests/transformations/tasklet_fusion_test.py (hunk 2 above)
Comment (tbennun) [reply tests/transformations/tasklet_fusion_test.py:24]: I'd just add another test

File: dace/transformation/dataflow/tasklet_fusion.py [hunk 6]
```diff
@@ -27,246 +29,197 @@ def visit_Name(self, node: ast.Name) -> Any:
...
 
+    def can_be_applied(self, graph: dace.SDFGState, expr_index: int, sdfg: dace.SDFG, permissive: bool = False) -> bool:
         t1 = self.t1
+        data = self.data if self.expr_index == 0 else None
         t2 = self.t2
 
-        # Tasklets must be of the same language
-        if t1.language != t2.language:
+        # Both Tasklets must be in Python
+        if t1.language is not Language.Python or t2.language is not Language.Python:
```
Comment (tbennun) [inline dace/transformation/dataflow/tasklet_fusion.py:134]: please do not use `is` operators for enums (only ==/!=)

File: dace/transformation/dataflow/tasklet_fusion.py [hunk 7]
```diff
@@ -27,246 +29,197 @@ def visit_Name(self, node: ast.Name) -> Any:
...
+        ]
 
+    def can_be_applied(self, graph: dace.SDFGState, expr_index: int, sdfg: dace.SDFG, permissive: bool = False) -> bool:
         t1 = self.t1
+        data = self.data if self.expr_index == 0 else None
         t2 = self.t2
 
-        # Tasklets must be of the same language
-        if t1.language != t2.language:
+        # Both Tasklets must be in Python
```
Comment (tbennun) [inline dace/transformation/dataflow/tasklet_fusion.py:133]: That's a bit sad. It was used in c2dace

**Review by tbennun — CHANGES_REQUESTED** [review]
Not there yet. Core functionality now missing

**Review by tbennun — APPROVED** [review]
LGTM if tests pass

**Review by orausch — COMMENTED** [review]
LGTM in general, but it would be nice to add the tests from DaCeML, most of them do not require other DaCeML features

**Pull request merged**